### DISCUSSION TITLE
Managing ACLs via KafkaUser with an external PKI

### DISCUSSION THREAD

**Discussion by LCaparelli** [strimzi/strimzi-kafka-operator#4829] [General]
Hey folks, thanks a lot for your work!

Our Kafka cluster is an AWS MSK instance and we were looking to use the Entity Operator to handle topics and ACLs. We're using mTLS for authentication and use Hashicorp Vault's PKI engine to manage our PKI.

In a nutshell: we want `KafkaUser`'s authorization stuff, but not the authentication stuff.

We're facing an issue when attempting to use the `KafkaUser` CR to manage the ACLs. Since all TLS/authentication stuff is handled by the Vault we leave `.spec.authentication` unspecified, preventing the Operator from making any certificate management procedure. However, since we're using mTLS auth, the ACL's user must be specified as the client's cert subject (e.g., "`CN=blabla,OU=foo,O=bar,L=etc,ST=etc,C=**`").

We believe that because we left `.spec.authentication` unspecified the Operator is creating the ACL with a principal that is simply the CR's name, when we need it to be the client cert's subject.

Now, we thought of working around this by simply naming the `KafkaUser` CR as the cert's subject, but unfortunately they don't produce a valid DNS name and get rejected by the API server. Is there a way to accomplish what we want without any changes to the existing code?

And if not, I believe a simple `.spec.username` overriding auth-based methods to define the username/principal would do the trick, but I'm not sure if that's something which you would like to support or if there's a better way to accomplish this. Any thoughts?

Thanks in advance for your attention and assistance. :-D

**Comment by scholzj** [comment]
When you leave out the `.spec.authentication` section, the operator assumes that the username is the name of the `KafkaUser` resource. And that will of course never be something like `CN=...`. So you cannot really use it for this scenario. All you can do is disable the User Operator completely and just use Kafka tools / APIs to manage the ACLs your self - that should be easy, just remove the `userOperator` section from the `Kafka` CR.

**Reply by LCaparelli** [reply]
@scholzj thanks for your response! As I said previously we're not using the cluster operator, just the topic and user operators.

So, from your response, there is no way to only use the authorization portion of the `KafkaUser`, orthogonal to authentication and this is not something which could potentially be supported, is that it?

If that's the case we'll probably  implement an operator to handle just authorization, though we would very much rather using this one.

**Reply by scholzj** [reply]
> As I said previously we're not using the cluster operator, just the topic and user operators.

Sorry, I missed that 🤦 

> So, from your response, there is no way to only use the authorization portion of the KafkaUser, orthogonal to authentication and this is not something which could potentially be supported, is that it?

Well, in theory there are two ways how this can be implemented:

* We can have a mode where you have some _noop_ TLS authentication to use the TLS style username for the ACLs. The limitation here would be that the username will be always `CN=<NameOfTheKafkaUserCR>`. E.g. `KafkaUser` resource named `my-user` will set the username in the ACLs or Quotas to `CN=my-user`. This would be easy to implement and support. But it will not be able to handle a user such as `CN=blabla,OU=foo,O=bar,L=etc,ST=etc`.

* The other option would be to have the username specified somewhere in the `.spec` section of the `KafkaUser` resource. But that is much harder to implement, because you will need to handle duplicate users, different user types etc. I think it will be also more complicated to have and support this long term. So I'm not sure how eager we would be to go for this (my personal view, other maintainers might see it differently).

**Reply by LCaparelli** [reply]
> Sorry, I missed that

No worries :-)

> We can have a mode where you have some noop TLS authentication to use the TLS style username for the ACLs. The limitation here would be that the username will be always CN=<NameOfTheKafkaUserCR>. E.g. KafkaUser resource named my-user will set the username in the ACLs or Quotas to CN=my-user. This would be easy to implement and support. But it will not be able to handle a user such as CN=blabla,OU=foo,O=bar,L=etc,ST=etc.

We could adjust the PKI config to generate certificates subjects using only the common name, I believe this would meet our use case requirements nicely. 

> The other option would be to have the username specified somewhere in the .spec section of the KafkaUser resource. But that is much harder to implement, because you will need to handle duplicate users, different user types etc. I think it will be also more complicated to have and support this long term. So I'm not sure how eager we would be to go for this (my personal view, other maintainers might see it differently).

Ah sorry, I didn't realize all the strings attached to this option when I suggested it.

My understanding of how the operator works is quite limited, but trying to chime in: would a `.spec.authorization.principalFormat` or similar field be simpler to support? Its default would depend on `.spec.authentication.type`, but would still allow for a user override. I suppose in this approach there's no need to handle duplicate users or anything like that, we simply change the representation of the username. Sort of a more restrained version of my initial suggestion.

**Reply by scholzj** [reply]
> We could adjust the PKI config to generate certificates subjects using only the common name, I believe this would meet our use case requirements nicely.

Ok ... one more caveat which I didn't specifically mentioned ... value of the common name would always need to be a valid Kubernetes resource to make it work like this.

> My understanding of how the operator works is quite limited, but trying to chime in: would a .spec.authorization.principalFormat or similar field be simpler to support? Its default would depend on .spec.authentication.type, but would still allow for a user override. I suppose in this approach there's no need to handle duplicate users or anything like that, we simply change the representation of the username. Sort of a more restrained version of my initial suggestion.

No. Because there are also quotas. So having something like 

```yaml
authentication:
  type: noop-tls
```

can be IMHO used to define the username for both ACLs and quotas and you do not need to have the format twice there. The format in the `authorization` section would IMHO also make it a but more complicated to understand all the different things which should be deleted.

**Reply by LCaparelli** [reply]
> Ok ... one more caveat which I didn't specifically mentioned ... value of the common name would always need to be a valid Kubernetes resource to make it work like this.

Yeah, this should be fine, we should be able to enforces all CNs are valid hostnames from the PKI config.

> can be IMHO used to define the username for both ACLs and quotas and you do not need to have the format twice there. The format in the authorization section would IMHO also make it a but more complicated to understand all the different things which should be deleted.

Yeah, that's reasonable, thanks for the discussion!

About the noop-tls, what should next steps be? Should I open an RFE? Wait for more contributors to participate here?

**Reply by scholzj** [reply]
Please open an enhancement issue. Do you want to give it a try and open a PR for it as well? Or should someone else look at it?

**Reply by LCaparelli** [reply]
> Ok ... one more caveat which I didn't specifically mentioned ... value of the common name would always need to be a valid Kubernetes resource to make it work like this.

Ok, just to clarify: valid Kubernetes resource **name**, right? This approach wouldn't create any Secrets.

**Reply by scholzj** [reply]
> Ok, just to clarify: valid Kubernetes resource **name**, right? This approach wouldn't create any Secrets.

No, it would not create any secrets. But you have to create the `KafkaUser` resource and the generated username will be based on its name. So you will be limited to `CN=<AnyValidKubernetesResourceName>` basically.

**Reply by LCaparelli** [reply]
@scholzj some of our requirements changed and moving forward we won't be using the Strimzi operator. :-(

Sorry for the hassle and thank you very much for your time.

In any case, if you still think this is a valid use case and a welcome feature I can still open the enhancement issue. I would love to give it a go with a PR, but Java isn't my jam and I'm not familiar with the codebase, would it be cool if came to you with questions about the design?

**Reply by scholzj** [reply]
No worries and thanks for letting me know. No need to open the issue if you don't need it anymore.
